[ResponseOps] move alert UUID generation from rule registry to the alerting framework

pmuellr opened the pull request:
resolves https://github.com/elastic/kibana/issues/142874

doc change: https://kibana_143489.docs-preview.app.elstc.co/guide/en/kibana/master/rule-action-variables.html , search on `alert.uuid`

## Summary

The alerting framework now generates an alert UUID for every alert it creates.  The UUID will be reused for alerts which continue to be active on subsequent runs, until the alert recovers.  When the same alert (alert instance id) becomes active again, a new UUID will be generated.  These UUIDs then identify a "span" of events for a single alert.

The rule registry plugin was already adding these UUIDs to it's own alerts-as-data indices, and that code has now been changed to make use of the new UUID the alerting framework generates.

- adds property in the rule task state `alertInstances[alertInstanceId].meta.uuid`; this is where the alert UUID is persisted across runs
- adds a new `Alert` method getUuid(): string` that can be used by rule executors to obtain the UUID of the alert they just retrieved from the factory; the rule registry uses this to get the UUID generated by the alerting framework
- for the event log, adds the property `kibana.alert.uuid` to `*-instance` event log events; this is the same field the rule registry writes into the alerts-as-data indices
- various changes to tests to accomodate new UUID data / methods
- migrates the UUID previous stored with lifecycle alerts in the alert state, via the rule registry *INTO* the new `meta.uuid` field in the existing alert state.

**deferred (follow-on issues created)**

- #144858 - Change rule registry to stop generating it's "wrapped" alert state, which was used to maintain the UUID persistence.  It's not needed any more, and likewise the "start date" accessor it provides can be added to the alerting framework, and use the field in the existing alert state.

- #144862 - Add support for rule executors to pass in explicit UUIDs instead of having the framework generate them.  This is needed to support rule registry "persistence" rule types.  The support in this PR is for "lifecycle" rules only.  Persistence rules generate the UUIDs themselves, and they aren't actually uuids, but sha's of data the alert is based on.  Presumably we'd add a new parameter to the alert factory method that returns alerts, or add a `setUuid(uuid: string)` function on `Alert` that rule executors could use.

### migration of UUID

Here is the current structure of the rule state, persisted as the rule's task state.  This is for a lifecycle alert:

<details>
<summary>lifecycle rule state, uuid field changes</summary>

```json
{
  "alertTypeState": {
    "wrapped": {
      "groups": [ "some-thing, metricbeat" ],
      "groupBy": [ "fields.eks_cluster_name", "agent.type" ],
      "filterQuery": ""
    },
    "trackedAlerts": {
      "some-thing, metricbeat": {
        "alertId": "some-thing, metricbeat",
        // migrated FROM below
        "alertUuid": "5cee8f86-5d0b-4d4a-92f7-a3c731b7ecc3", 
        "started": "2022-03-11T20:46:24.170Z"
      },
    }
  },
  "alertInstances": {
    "some-thing, metricbeat": {
      "state": {
        "start": "2022-03-11T20:46:24.383Z",
        "duration": 16913558599000000
      },
      "meta": {
        // migrated TO below
        "uuid": "5cee8f86-5d0b-4d4a-92f7-a3c731b7ecc3",
        "lastScheduledActions": {
          "group": "metrics.threshold.fired",
          "date": "2022-03-11T20:46:24.390Z"
        }
      }
    },
  },
  "previousStartedAt": "2022-09-23T14:59:02.346Z"
}
```

In addition to moving the uuid field, the migration will also create a new uuid for non-lifecycle events, and in cases where an alert is listed in the rule registry's `trackedAlerts`, but not in the framework's `alertInstances`, it should be added there.

### Checklist

Delete any items that are not applicable to this PR.

- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

pmuellr committed:
[ResponseOps] move alert UUID generation from rule registry to alerting framework

resolves https://github.com/elastic/kibana/issues/142874

pmuellr commented on x-pack/plugins/rule_registry/server/utils/create_lifecycle_executor.ts:
```diff
@@ -170,17 +169,21 @@ export const createLifecycleExecutor =
       alertWithLifecycle: ({ id, fields }) => {
         currentAlerts[id] = fields;
-        return alertFactory.create(id);
+        const alert = alertFactory.create(id);
```
this is where the rule registry gets the UUID created by the alerting framework, instead of generating it here.

pmuellr commented on x-pack/plugins/rule_registry/server/utils/create_lifecycle_executor.ts:
```diff
@@ -230,22 +233,30 @@ export const createLifecycleExecutor =
           ? state.trackedAlerts[alertId]
           : {
-              alertUuid: newAlertUuids[alertId] || v4(),
+              alertUuid: alertUuidMap.get(alertId),
```
this is where we update the alert uuid that the rule registry currently tracks, which is no longer actively used.  We'll get rid of this in https://github.com/elastic/kibana/issues/144858

pmuellr commented on x-pack/plugins/rule_registry/server/utils/create_lifecycle_executor.ts:
```diff
@@ -230,22 +233,30 @@ export const createLifecycleExecutor =
               started: commonRuleFields[TIMESTAMP],
             };
 
+        const alertUuid = alertUuidCalc ?? v4();
```
This shouldn't happen, but if for some reason the alert uuid calculation does not work, revert to having the rule registry generate it and log a message that it did.

pmuellr commented on x-pack/plugins/security_solution/server/lib/detection_engine/signals/preview/alert_instance_factory_stub.ts:
```diff
@@ -49,4 +50,7 @@ export const alertInstanceFactoryStub = <
   hasContext() {
     return false;
   },
+  getUuid() {
```
The "preview" code somehow manages to run the rule executor locally, and so needs to have stub versions of everything available, typing wise.

pmuellr commented on x-pack/test/rule_registry/common/lib/helpers/cleanup_target_indices.ts:
```diff
@@ -17,7 +17,7 @@ export const cleanupTargetIndices = async (getService: GetService, user: User, s
     const aliasMap = await es.indices.getAlias({
       name: targetIndices,
       allow_no_indices: true,
-      expand_wildcards: 'open',
```
I think this was the main reason the rule registry test below was skipped.  That test was the best place for uuid tests, so managed to get it running again \o/

pmuellr replied:
hmmm ... I'm not sure `all` is right though either.  I wonder if `open` was used to prevent some of the RR init problems with non-hot-tiered indices (eg, `partial-*`), and so using `all` would make those scenarios more likely ...

pmuellr commented on x-pack/test/rule_registry/spaces_only/tests/trial/create_rule.ts:
```diff
@@ -195,9 +196,23 @@ export default function registryRulesApiTest({ getService }: FtrProviderContext)
+          expect(typeof alertUuid).to.be('string');
+          expect(typeof executionUuid).to.be('string');
+
+          await checkEventLogAlertUuids(
```
this is where we check that the uuid in the AaD indices and the event log are the same

elasticmachine commented:
Pinging @elastic/response-ops (Team:ResponseOps)

pmuellr commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into alerting/alert-uuid

elasticmachine commented:
Pinging @elastic/apm-ui (Team:APM)

pmuellr committed:
add retries for a function test

CoenWarmer commented on x-pack/plugins/alerting/README.md:
```diff
@@ -766,6 +766,7 @@ This factory returns an instance of `Alert`. The `Alert` class has the following
 
 |Method|Description|
 |---|---|
+|getUuid()|Get the UUID of the alert.|
```
Should line 786 also include `alertUuid` in the list of `Templating Actions`?

pmuellr replied:
Rather than update a probably already incomplete list, I added a link to the public doc, which I believe we keep up-to-date a little better.

CoenWarmer commented on x-pack/plugins/alerting/common/alert_summary.ts:
```diff
@@ -32,6 +32,7 @@ export interface AlertSummary {
 }
 
 export interface AlertStatus {
+  uuid?: string;
```
In which case will `uuid` be optional?

pmuellr replied:
This is kind of a funky data structure.  It sometimes contains a "muted" indicator for alerts which are not currently active, and so don't also have an `actionGroupId`, and won't have a `uuid`.  Beyond that, existing rule's task state when migrated to this Kibana version will not have the `uuid` field set, so I wanted to account for that, typing-wise.

CoenWarmer commented on x-pack/plugins/alerting/server/alert/alert.ts:
```diff
@@ -33,7 +35,13 @@ export type PublicAlert<
+  | 'setContext'
+  | 'getContext'
+  | 'hasContext'
+  | 'getUuid'
```
Nit: alphabetize for less noise

CoenWarmer commented on x-pack/plugins/alerting/server/lib/alert_summary_from_event_log.ts:
```diff
@@ -149,10 +150,15 @@ export function alertSummaryFromEventLog(params: AlertSummaryFromEventLogParams)
+function getAlertStatus(
+  alerts: Map<string, AlertStatus>,
+  alertId: string,
+  alertUuid?: string
```
Nit: I prefer `alertUuid: string | undefined` as it is a bit more explicit.

pmuellr replied:
There are other uses of this function which do not use the uuid, so changing to `string | undefined` would mean those uses would explicitly have to pass `undefined`.  So, seems like "optional" with `?` works better here.

CoenWarmer commented on x-pack/plugins/alerting/server/lib/alerting_event_logger/alerting_event_logger.ts:
```diff
@@ -45,6 +45,7 @@ interface DoneOpts {
 interface AlertOpts {
   action: string;
   id: string;
+  uuid?: string;
```
Same as earlier, I'm wondering in which case `uuid` would not be available. It might be `string` or `undefined` though.

pmuellr replied:
Changing this one to remove the optional, since the uuid should always be a string when this is used.

CoenWarmer reviewed (approved):
Looks great, excited to use this! 

Some minor comments.

ersin-erdal reviewed (approved):
LGTM,
Tested locally and was able to see the uuid in AAD, EventLog and the alert object.

marshallmain reviewed (approved):
Alerts area changes LGTM

ymao1 commented:
@pmuellr When I have an existing "active" lifecycle alert with an existing UUID, is it expected that the UUIDs won't match with this PR?

For example, if I have an existing "active" alert prior to this PR with `uuid-1` and then I switch to this PR branch. The rule runs and generates the active alert again. It looks like the existing AAD document is updated but the `kibana.alert.uuid` in that doesn't match the newly generated `kibana.alert.uuid` in the event log `active-instance` doc. Likely the alert would have to recover and then become active again to get a new UUID from the framework. Does that affect anything with flapping down the line if those don't match up?

pmuellr commented:
> When I have an existing "active" lifecycle alert with an existing UUID, is it expected that the UUIDs won't match with this PR?

It was intended for this case to be handled; the idea for lifecycle rules was that it would always use the old "wrapped" version if found, so would be the same over a migration.

I know I tested it manually at one point, but thinking now I didn't get a proper test in the lifecycle tests.  Broke it fixing something else, prolly :-) . Thanks, will look into it!

pmuellr commented:
> When I have an existing "active" lifecycle alert with an existing UUID, is it expected that the UUIDs won't match with this PR?

Dang.  So, in the original design, where we would migrate the "wrapped" lifecycle executor structure away, we would have set the new `uuid` field in the alert task state to the one from the wrapped structure.  But that part got deferred in https://github.com/elastic/kibana/issues/144858 .  And there's not really a "clean" way to do this, without either the migration OR we implement https://github.com/elastic/kibana/issues/144862 and allow the lifecycle executor to pass in the old uuid when creating the alert.  

I think #144862 may not be too bad to implement, #144858 is going to be painful though - I'd like to keep that migration code in a smaller PR.  Problem with #144862 is that the "fix" to the lifecycle rule would be temporary till we get rid of the wrapped structure, which at that point will align currently active alert UUIDs AND not require the lifecycle executor to generated UUIDs at all (will use the one from the alerting framework).  Weird, but seems fine.  And will need to update that issue to point out the additional work that will be needed.

pmuellr commented on x-pack/plugins/alerting/server/alert/alert.ts:
```diff
@@ -46,18 +54,24 @@ export class Alert<
   private state: State;
   private context: Context;
   private readonly id: string;
+  private readonly uuid: string;
```
Defensively storing the `uuid` in the alert, as we occaisonally mutate `meta`, so to make sure we don't get an unexpected uuid from a meta mutation, store in the alert, and copy it into meta when the alert is created.

pmuellr commented on x-pack/test/alerting_api_integration/spaces_only/tests/alerting/event_log.ts:
```diff
@@ -17,6 +17,12 @@ import {
 } from '../../../common/lib';
 import { FtrProviderContext } from '../../../common/ftr_provider_context';
 
+const InstanceActions = new Set<string | undefined>([
```
I think we can import these from alerting, instead of copying ...

pmuellr commented:
Got asked yesterday "why haven't we merged this?".  The original thinking that this would need to be done before adding the flapping logic - probably involving getting rid of the "wrapped" rule task state as well.  Instead, we went down a path of flapping code for RR (rule registry) similar to what we did for the alerting framework.  So ... this work could be deferred.

Will be doing a bit of planning to see where we all with all this stuff re: flapping but mainly FAAD (framework alerts as data). 

Beyond that, there are some PR comments that still need to be addressed.  And I believe https://github.com/elastic/kibana/pull/143489#issuecomment-1315458981 indicates that we really need to add a migration here, to move the uuid from the rule registry rule task state (in the "wrapped structure") into the alerting framework section, so that the same UUIDs are between RR and AF (alerting framework), after a migration.

pmuellr committed:
merge main and fix _some_ stuff

pmuellr committed:
Merge branch 'main' into alerting/alert-uuid

pmuellr committed:
merge main and fix conflicts

pmuellr committed:
add back some doc I removed when fixing merge conflicts

pmuellr committed:
fix jest tests after merging main

pmuellr committed:
fix some FT, start on migrating UUID out of RR alerts into framework meta

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

pmuellr commented:
Converted this to a draft while I add the migration of the uuid in the rule registry wrapper to the framework fields.

pmuellr committed:
add a zx sctipt to get task docs to write tests agains

pmuellr committed:
finished jest tests for migration

pmuellr committed:
get tests working again

pmuellr committed:
fix file name issue

pmuellr commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into alerting/alert-uuid

pmuellr commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into alerting/alert-uuid

pmuellr commented:
buildkite test this

pmuellr commented:
Failing tests all seem flaky to me, so ... trying again - and may try tonight if it still fails again.

pmuellr committed:
remove rendundant data from test

pmuellr committed:
move tm-shared types to a new package

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/generate codeowners'

pmuellr commented:
buildkite test this

pmuellr committed:
another attempt at building cross-package type package

pmuellr committed:
fix jest config to point to right root, and fix license in package.json

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/generate codeowners'

pmuellr committed:
merge main, fix conflicts

pmuellr committed:
Merge remote-tracking branch 'pmuellr/alerting/alert-uuid' into alerting/alert-uuid

to deal with the auto-commit

pmuellr committed:
fix tests failing after main merge

pmuellr committed:
rearrange package re-export in hopes of fixing lint error

kibanamachine committed:
Merge branch 'main' into alerting/alert-uuid

pmuellr committed:
various cleanups; function tests next

pmuellr committed:
add function tests

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/lint_ts_projects --fix'

pmuellr committed:
fix UUID being removed when storing recovered alerts for flapping

pmuellr committed:
Merge remote-tracking branch 'pmuellr/alerting/alert-uuid' into alerting/alert-uuid

pmuellr committed:
merge main, fix merge conflict in x-pack/test/tsconfig.json

pmuellr committed:
add additional function test

pmuellr commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into alerting/alert-uuid

pmuellr committed:
fix doc

pmuellr commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into alerting/alert-uuid

pmuellr commented on .github/CODEOWNERS:
```diff
@@ -18,6 +18,7 @@ x-pack/test/alerting_api_integration/common/plugins/alerts @elastic/response-ops
 x-pack/examples/alerting_example @elastic/response-ops
 x-pack/test/functional_with_es_ssl/plugins/alerts @elastic/response-ops
 x-pack/plugins/alerting @elastic/response-ops
+x-pack/packages/kbn-alerting-state-types @elastic/response-ops
```
This is a new package added in this PR, to hold alerting types used by task manager.  More below.

pmuellr commented on x-pack/packages/kbn-alerting-state-types/README.md:
```diff
@@ -0,0 +1,8 @@
+# @kbn/alerting-state-types
```
This package was added so that we can use the same types used in alerting and rule_registry, for rule task state, within task manager, for migration reasons. Alerting and rule_registry now import this package to get those types, and also task_manager.

No changes were made to the types. The files are not arranged in a super-logical fashion - I left them pretty much in the same files as where they originally came from, to make it easier to diff.

pmuellr commented on x-pack/plugins/task_manager/server/saved_objects/migrations.ts:
```diff
@@ -13,16 +14,22 @@ import {
+
 import { REMOVED_TYPES } from '../task_type_dictionary';
-import { ConcreteTaskInstance, TaskStatus } from '../task';
+import { SerializedConcreteTaskInstance, TaskStatus } from '../task';
```
We've been using the wrong type, all these years!  The `ConcreteTaskInstance` type has JS types - for example, `Date` objects for dates.  The `SerializedConcreteTaskInstance` has the serialized version, so ISO date strings for dates.  The data passed here is the _serialized_ version, not the _JS_ version.

We need the Serialized version for the new migration, so I changed the types of all the migrations.  The existing migration didn't have to deal with fields that were different between the two, so it was safe to change it.

pmuellr commented:
Hey folks!  👋🏻 

You may remember this PR from last year!  We didn't merge it, as we realized we really wanted a migration so that the alerting framework UUIDs would match the UUIDs the rule registry is generating, for alerts active at the time of the upgrade.

And then we parked the PR for a bit, then X-mas, then ...

But ... **_we're back, baby_**!  I don't think much has changed since the reviews were first made, except for the following:

- added a migration of task manager document for rules, to add UUIDs for all alerts stored in task state. For rules using the rule_registry, life cycle rules will keep the same UUIDs they had before the upgrade, and these become the UUIDs for the alerts in the alerting framework task storage.  We still don't do anything with persistent rules - they will have different alert uuids in the alerting framework, and in the alerts-as-data indices - see https://github.com/elastic/kibana/issues/144862 
- had to move types from alerting and rule_registry - the task doc shapes - into a package that task manager could import, so that we're using the same types everywhere
- because of the migration, we now no longer have to assume the uuid field will be undefined, it should always be a string; so there were a few places we removed `| undefined` on the typing of the uuid value

ymao1 reviewed (approved):
LGTM! Verified alert UUIDs are generated for new alerts and migrated for existing lifecycle alerts. Amazing test coverage :)

pmuellr committed:
merge main and fix simple merge conflict

lcawl commented on docs/user/alerting/action-variables.asciidoc:
```diff
@@ -100,6 +100,7 @@ If the rule's action frequency is not a summary of alerts, it passes the followi
 `alert.actionSubgroup`:: The action subgroup of the alert that scheduled the action.
 `alert.flapping`:: A flag on the alert that indicates whether the alert status is changing repeatedly.
 `alert.id`:: The ID of the alert that scheduled the action.
+`alert.uuid`:: The UUID of the alert.  The UUID value will remain unchanged while the alert is active, during subsequent runs.  preview:[]
```
If necessary, these changes can be in a subsequent PR, but here's my suggestion:
```suggestion
`alert.uuid`:: A universally unique identifier for the alert. While the alert is active, the UUID value remains unchanged each time the rule runs.  preview:[]
```

lcawl reviewed (approved):
One docs suggestion, otherwise text LGTM

pmuellr committed:
fix test broken after main merge, fix doc per PR comment

pmuellr commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into alerting/alert-uuid
